server/storage/database_test.go

```diff
@@ -257,53 +258,126 @@ func TestSQLDelete(t *testing.T) {
 
-	var rows []TimestampKey
+	var rows []Key
 	query := gormDB.Select("ID, Gun, Cipher, Public").Find(&rows)
 	assert.NoError(t, query.Error)
 
-	expected := TimestampKey{Gun: "testGUN", Cipher: "testCipher",
+	expected := Key{Gun: "testGUN", Cipher: "testCipher",
 		Public: []byte("1")}
 	expected.Model = gorm.Model{ID: 1}
 
-	assert.Equal(t, []TimestampKey{expected}, rows)
+	assert.Equal(t, []Key{expected}, rows)
+
+	dbStore.DB.Close()
+}
+
+func TestSQLSetKeyMultipleRoles(t *testing.T) {
+	tempBaseDir, err := ioutil.TempDir("", "notary-test-")
+	gormDB, dbStore := SetUpSQLite(t, tempBaseDir)
+	defer os.RemoveAll(tempBaseDir)
+
+	err = dbStore.SetKey("testGUN", data.CanonicalTimestampRole, "testCipher", []byte("1"))
+	assert.NoError(t, err, "Inserting timestamp into empty DB should succeed")
+
+	err = dbStore.SetKey("testGUN", data.CanonicalSnapshotRole, "testCipher", []byte("1"))
+	assert.NoError(t, err, "Inserting snapshot key into DB with timestamp key should succeed")
+
+	var rows []Key
+	query := gormDB.Select("ID, Gun, Role, Cipher, Public").Find(&rows)
+	assert.NoError(t, query.Error)
+
+	expectedTS := Key{Gun: "testGUN", Role: "timestamp", Cipher: "testCipher",
+		Public: []byte("1")}
+	expectedTS.Model = gorm.Model{ID: 1}
+
+	expectedSN := Key{Gun: "testGUN", Role: "snapshot", Cipher: "testCipher",
+		Public: []byte("1")}
+	expectedSN.Model = gorm.Model{ID: 2}
+
+	assert.Equal(t, []Key{expectedTS, expectedSN}, rows)
+
+	dbStore.DB.Close()
+}
+
+func TestSQLSetKeyMultipleGuns(t *testing.T) {
+	tempBaseDir, err := ioutil.TempDir("", "notary-test-")
+	gormDB, dbStore := SetUpSQLite(t, tempBaseDir)
+	defer os.RemoveAll(tempBaseDir)
+
+	err = dbStore.SetKey("testGUN", data.CanonicalTimestampRole, "testCipher", []byte("1"))
+	assert.NoError(t, err, "Inserting timestamp into empty DB should succeed")
+
+	err = dbStore.SetKey("testAnotherGUN", data.CanonicalTimestampRole, "testCipher", []byte("1"))
+	assert.NoError(t, err, "Inserting snapshot key into DB with timestamp key should succeed")
+
+	var rows []Key
+	query := gormDB.Select("ID, Gun, Role, Cipher, Public").Find(&rows)
+	assert.NoError(t, query.Error)
+
+	expected1 := Key{Gun: "testGUN", Role: "timestamp", Cipher: "testCipher",
+		Public: []byte("1")}
+	expected1.Model = gorm.Model{ID: 1}
+
+	expected2 := Key{Gun: "testAnotherGUN", Role: "timestamp", Cipher: "testCipher",
+		Public: []byte("1")}
+	expected2.Model = gorm.Model{ID: 2}
+
+	assert.Equal(t, []Key{expected1, expected2}, rows)
+
+	dbStore.DB.Close()
+}
+
+func TestSQLSetKeySameRoleGun(t *testing.T) {
+	tempBaseDir, err := ioutil.TempDir("", "notary-test-")
+	_, dbStore := SetUpSQLite(t, tempBaseDir)
+	defer os.RemoveAll(tempBaseDir)
+
+	err = dbStore.SetKey("testGUN", data.CanonicalTimestampRole, "testCipher", []byte("1"))
+	assert.NoError(t, err, "Inserting timestamp into empty DB should succeed")
+
+	err = dbStore.SetKey("testGUN", data.CanonicalTimestampRole, "testCipher", []byte("1"))
```

Comment:
Thanks!  Apologies, I should have added a test for this one before.  :)  Could we change this to `[]byte("2")`?  If it's `[]byte("1")`, then the uniqueness constraint might be on the combination of "gun", "role", and "key", whereas we want to test that the uniqueness constraint is on "gun" and "role" only (you can't add another key for the same gun and role)